core/drivers/crypto/stm32/stm32_pka.c

```diff
@@ -0,0 +1,1734 @@
+				    curve_def[cid].b.val, curve_def[cid].b.size,
+				    eo_nbw);
+		if (res)
+			return res;
+	}
+
+	if (pka_ram[op][PRIME_N]) {
+		res = write_eo_data(base + pka_ram[op][PRIME_N],
+				    curve_def[cid].n.val, curve_def[cid].n.size,
+				    eo_nbw);
+		if (res)
+			return res;
+	}
+
+	res = write_eo_data(base + pka_ram[op][VAL_P], curve_def[cid].p.val,
+			    curve_def[cid].p.size, eo_nbw);
+	if (res)
+		return res;
+
+	if (pka_ram[op][GPOINT_X]) {
+		res = write_eo_data(base + pka_ram[op][GPOINT_X],
+				    curve_def[cid].g.x.val,
+				    curve_def[cid].g.x.size, eo_nbw);
+		if (res)
+			return res;
+	}
+
+	if (pka_ram[op][GPOINT_Y]) {
+		res = write_eo_data(base + pka_ram[op][GPOINT_Y],
+				    curve_def[cid].g.y.val,
+				    curve_def[cid].g.y.size, eo_nbw);
+		if (res)
+			return res;
+	}
+
+	return TEE_SUCCESS;
+}
+
+/**
+ * @brief Check if BigInt stored in d is 0
+ *
+ * @retval: true: if data represents a 0 value (i.e. all bytes == 0)
+ *          false: if data represents a non-zero value.
+ */
+static bool is_zero(const struct stm32_pka_bn *d)
+{
+	unsigned int i = U(0);
+
+	assert(d);
+
+	for (i = U(0); i < d->size; i++)
+		if (d->val[i] != U(0))
+			return false;
+
+	return true;
+}
+
+/**
+ * @brief Compare two BigInt:
+ *
+ * @retval: true if a < b
+ *          false if a >= b
+ */
+static bool is_smaller(const struct stm32_pka_bn *a,
+		       const struct stm32_pka_bn *b)
+{
+	unsigned int i = MAX(a->size, b->size) + 1;
+
+	for (; i != U(1); i--) {
+		uint8_t _a = U(0);
+		uint8_t _b = U(0);
+
+		if (a->size >= i)
+			_a = a->val[a->size - i];
+		if (b->size >= i)
+			_b = b->val[b->size - i];
+
+		if (_a < _b)
+			return true;
+		if (_a > _b)
+			return false;
+	}
+
+	return false;
+}
+
+TEE_Result stm32_pka_get_max_size(size_t *bytes, size_t *bits,
+				  const enum stm32_pka_curve_id cid)
+{
+	if (cid < 0 || cid >= PKA_LAST_CID)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	if (bits)
+		*bits = curve_def[cid].n_len;
+
+	if (bytes)
+		*bytes = curve_def[cid].n.size;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result stm32_pka_compute_r2modn_ret(const vaddr_t base,
+					       struct stm32_pka_bn *v,
+					       const unsigned int eo_nbw)
+{
+	uint32_t sr = U(0);
+
+	sr = io_read32(base + _PKA_SR);
+	if ((sr & (_PKA_IT_OPERR | _PKA_IT_ADDRERR | _PKA_IT_RAMERR)) != 0) {
+		EMSG("Detected error(s): %s%s%s",
+		     (sr & _PKA_IT_OPERR) ? "Operation " : "",
+		     (sr & _PKA_IT_ADDRERR) ? "Address " : "",
+		     (sr & _PKA_IT_RAMERR) ? "RAM" : "");
+		return TEE_ERROR_SECURITY;
+	}
+
+	return read_eo_data(base + _PKA_RAM_R2MODN_OUT, v->val, v->size,
+			    eo_nbw);
+}
+
+TEE_Result stm32_pka_compute_montgomery(const struct stm32_pka_bn *n,
+					const size_t n_len,
+					struct stm32_pka_bn *r2modn)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	vaddr_t base = pka_pdata.base;
+	unsigned int eo_nbw = OP_NBW_FROM_LEN(n_len);
+
+	if (!n_len || !n || !r2modn)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	mutex_lock(pka_pdata.lock);
```

Comment:
Would'nt you prefer to use `mutex_lock(&pka_lock)` straight here and everywhere the mutex is locked/unlocked?